File: frontend-new/src/types/drift.ts
export interface DriftMilestone {
  horizon_hours: number;
  timestamp: string;
  latitude: number;
  longitude: number;
  distance_km: number;
  speed_ms: number;
  heading_deg: number;
  uncertainty_radius_km: number;
}

export interface DriftForecastSummary {
  forecast_id: string;
  contact_id?: string | null;
  initial_latitude: number;
  initial_longitude: number;
  initial_timestamp: string;
  depth_m: number;
  model_name: string;
  model_version: string;
  champion_model: string;
  execution_mode: string;
  hotspot_score: number;
  hotspot_evidence: Record<string, any>;
  milestones: Record<string, DriftMilestone>;
  created_at: string;
}

export interface DriftForecastDetail extends DriftForecastSummary {
  geojson: any;
}

export interface OceanDataStatus {
  copernicus: {
    available: boolean;
    dataset_product_id?: string;
    dataset_id?: string;
    drift_mode?: string;
    time_range?: string[];
    depth_levels?: number[];
    latitude_range?: number[];
    longitude_range?: number[];
    variables_available: string[];
    variables_count?: number;
    temporal_resolution?: string;
    spatial_resolution_deg?: number;
    last_update?: string;
    error?: string;
  };
  incois: {
    available: boolean;
    status: string;
    source_mode?: string;
    endpoint_url?: string;
    dataset_name?: string;
    product_family?: string;
    columns: string[];
    variables: string[];
    total_records?: number;
    spatial_coverage?: string;
    last_update?: string;
    error?: string;
  };
  summary: {
    all_sources_ready: boolean;
    drift_mode: string;
    copernicus_variables_count: number;
    incois_status: string;
    missing_variables: string[];
  };
  last_checked: string;
}

export interface DriftModelInfo {
  model_name: string;
  model_class: string;
  version: string;
  status: string;
  description: string;
  metrics: {
    '24h'?: { mae_km: number; rmse_km: number; median_km: number };
    '48h'?: { mae_km: number; rmse_km: number; median_km: number };
    '72h'?: { mae_km: number; rmse_km: number; median_km: number };
    status: string;
  };
}

File: backend/app/schemas/drift.py
"""
Pydantic Schemas for Drift Forecasting and Ocean Intelligence.
"""

from typing import Optional, Dict, Any, List
from datetime import datetime
from pydantic import BaseModel, Field


class DriftPredictionRequest(BaseModel):
    contact_id: Optional[str] = Field(default=None, description="Contact ID to forecast drift for")
    latitude: Optional[float] = Field(default=None, description="Initial latitude if contact_id not provided")
    longitude: Optional[float] = Field(default=None, description="Initial longitude if contact_id not provided")
    timestamp: Optional[datetime] = Field(default=None, description="Detection timestamp (UTC)")
    horizon_hours: int = Field(default=72, ge=1, le=168, description="Forecast horizon in hours (default 72h)")
    model_preference: Optional[str] = Field(default="CHAMPION", description="Model: CHAMPION | PHYSICS | PHYSICS_GRU | PHYSICS_LSTM")


class DriftMilestone(BaseModel):
    horizon_hours: float
    timestamp: str
    latitude: float
    longitude: float
    distance_km: float
    speed_ms: float
    heading_deg: float
    uncertainty_radius_km: float


class DriftForecastSummary(BaseModel):
    forecast_id: str
    contact_id: Optional[str] = None
    initial_latitude: float
    initial_longitude: float
    initial_timestamp: str
    depth_m: float
    model_name: str
    model_version: str
    champion_model: str
    execution_mode: str
    hotspot_score: float
    hotspot_evidence: Dict[str, Any] = Field(default_factory=dict)
    milestones: Dict[str, Any] = Field(default_factory=dict)
    created_at: str


class DriftForecastDetail(DriftForecastSummary):
    geojson: Dict[str, Any]


class OceanDataStatusResponse(BaseModel):
    copernicus: Dict[str, Any]
    incois: Dict[str, Any]
    summary: Dict[str, Any]
    last_checked: str


class DriftMetricsResponse(BaseModel):
    champion_model: str
    comparison_timestamp: str
    models: List[Dict[str, Any]]
